pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/MutableSegmentImpl.java

```diff
@@ -558,15 +560,39 @@ private RecordInfo getRecordInfo(GenericRow row, int docId) {
     PrimaryKey primaryKey = row.getPrimaryKey(_schema.getPrimaryKeyColumns());
 
     if (isUpsertEnabled()) {
-      Object upsertComparisonValue = row.getValue(_upsertComparisonColumn);
-      Preconditions.checkState(upsertComparisonValue instanceof Comparable,
-          "Upsert comparison column: %s must be comparable", _upsertComparisonColumn);
-      return new RecordInfo(primaryKey, docId, (Comparable) upsertComparisonValue);
+      if (_upsertComparisonColumns.size() > 1) {
+        return multiComparisonRecordInfo(primaryKey, docId, row);
+      }
+      Comparable comparisonValue = (Comparable) row.getValue(_upsertComparisonColumns.get(0));
+      return new RecordInfo(primaryKey, docId, comparisonValue);
     }
 
     return new RecordInfo(primaryKey, docId, null);
   }
 
+  private RecordInfo multiComparisonRecordInfo(PrimaryKey primaryKey, int docId, GenericRow row) {
+    Comparable[] comparisonColumns = new Comparable[_upsertComparisonColumns.size()];
+
+    int i = -1;
+    for (String columnName : _upsertComparisonColumns) {
+      i++;
+      Object comparisonValue = row.getValue(columnName);
+
+      Preconditions.checkState(comparisonValue instanceof Comparable,
+          "Upsert comparison column: %s must be comparable", columnName);
+
+      if (row.isNullValue(columnName)) {
+        // Inbound records may only have exactly 1 non-null value in one of the comparison column i.e. comparison
+        // columns are mutually exclusive
+        comparisonColumns[i] = null;
+        continue;
+      }
+
+      comparisonColumns[i] = (Comparable) comparisonValue;
```

Comment:
Check `null` first, then read value only if it is not `null`
```suggestion
      if (!row.isNullValue(columnName)) {
        Object comparisonValue = row.getValue(columnName);
        ...
        comparisonColumns[i] = (Comparable) comparisonValue;
      }
```